compiler/next/lib/resolution/resolution-types.cpp

```diff
@@ -17,11 +17,235 @@
  * limitations under the License.
  */
 
+#include "chpl/queries/update-functions.h"
 #include "chpl/resolution/resolution-types.h"
+#include "chpl/uast/Builder.h"
+#include "chpl/uast/Formal.h"
 
 namespace chpl {
 namespace resolution {
 
+using namespace uast;
+
+void ResolutionResultByPostorderID::setupForSymbol(const ASTNode* ast) {
+  assert(Builder::astTagIndicatesNewIdScope(ast->tag()));
+  vec.resize(ast->id().numContainedChildren());
+
+  symbolId = ast->id();
+}
+void ResolutionResultByPostorderID::setupForSignature(const Function* func) {
+  int bodyPostorder = 0;
+  if (func && func->body())
+    bodyPostorder = func->body()->id().postOrderId();
+  assert(0 <= bodyPostorder);
+  vec.resize(bodyPostorder);
+
+  symbolId = func->id();
+}
+void ResolutionResultByPostorderID::setupForFunction(const Function* func) {
+  setupForSymbol(func);
+}
+
+bool ResolutionResultByPostorderID::update(ResolutionResultByPostorderID& keep,
+                                           ResolutionResultByPostorderID& addin)
+{
+  return defaultUpdateVec(keep.vec, addin.vec);
+}
+
+bool FormalActualMap::computeAlignment(const UntypedFnSignature* untyped,
+                                       const TypedFnSignature* typed,
+                                       const CallInfo& call) {
+
+  // untyped must be provided but typed can be nullptr.
+  assert(untyped);
+
+  // create the mapping handling named and default arguments
+
+  // clear the current mapping
+  byFormalIdx.empty();
+  actualIdxToFormalIdx.empty();
+  mappingIsValid = false;
+  failingActualIdx = -1;
+
+  // allocate space in the arrays
+  byFormalIdx.resize(untyped->formals.size());
+  actualIdxToFormalIdx.resize(call.actuals.size());
+
+  // initialize the FormalActual parts from the Formals
+  size_t formalIdx = 0;
+  for (const Formal* formal : untyped->formals) {
+    FormalActual& entry = byFormalIdx[formalIdx];
+    entry.formal = formal;
+    if (typed) {
+      entry.formalType = typed->formalTypes[formalIdx];
+    }
+    entry.hasActual = false;
+    entry.actualIdx = -1;
+
+    formalIdx++;
+  }
+
+  // Match named actuals against formal names in the function signature.
+  // Record successful matches in actualIdxToFormalIdx.
+
+  size_t actualIdx = 0;
+  for (const CallInfoActual& actual : call.actuals) {
+    if (!actual.byName.isEmpty()) {
+      bool match = false;
+      int formalIdx = 0;
+      for (const Formal* formal : untyped->formals) {
+        if (actual.byName == formal->name()) {
+          match = true;
+          FormalActual& entry = byFormalIdx[formalIdx];
+          entry.hasActual = true;
+          entry.actualIdx = actualIdx;
+          entry.actualType = actual.type;
+          actualIdxToFormalIdx[actualIdx] = formalIdx;
```

Comment:
Not something to handle now necessarily, but is it possible this won't catch accidentally naming the same argument twice in the same call?

Reply:
This code is a transliteration of the same logic in the production compiler, so I would expect it to handle it. However, we'll find out when we get to running `start_test` (as I'm pretty sure we have a test of that, somewhere...)

Reply:
> This code is a transliteration of the same logic in the production compiler, so I would expect it to handle it. 

I don't know that that's a fair assumption to make - I don't think you copied all the pieces, right?  I don't see logic about double checking there isn't a duplicate named argument in this function - I think that happened in the wrapper generation and I don't remember seeing the wrapper code added to this PR.

But we will no doubt catch it when running start_test (I'm mostly trying to save you time tracking down why those tests would break - you're going to get a lot of failures at that point for things we're not remembering at the moment, having a category of things we know aren't handled before then will only save time diagnosing)

Reply:
I will mention it in the Future Work section of the PR message.